File: android/src/main/kotlin/com/example/flutter_crypto_algorithm/CryptoHelper.kt
package com.example.flutter_crypto_algorithm

import android.util.Base64
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.*

class CryptoHelper {
    enum class CryptoType {
        ENCRYPT,
        DECRYPT,
    }

    object CONSTANTS {
        const val ERR_VALID_KEY = "Invalid key or corrupted data"
        const val ERR_INCORRECT_BLOCK_SIZE = "Incorrect block size"
        const val ERR_UNEXPECTED_ERROR = "Unexpected error"
        const val KEY_256 = 256
    }

    object AES {
        const val ALGORITHM = "AES"
        const val TRANSFORMATION = "AES/CBC/PKCS5PADDING"
        val EMPTY_IV_SPEC = IvParameterSpec(ByteArray(16) { 0x00.toByte() })
        val EMPTY_SECRET_KEY = SecretKeySpec(ByteArray(32) { 0x00 }, "AES")
    }

    private fun genExactlyCharacterLengthKey(key: String, length: Int): String {
        val bytes = key.toByteArray(Charsets.UTF_8)
        val hexString = bytes.joinToString("") { "%02x".format(it) }
        val hexLength = hexString.take(length).padEnd(length, '0');
        return hexLength;
    }

    fun genSecretKey(algorithmType: String, secretKey: String?): SecretKey? {
        return when (algorithmType) {
            AES.ALGORITHM -> {
                if (secretKey != null) {
                    val transform = genExactlyCharacterLengthKey(secretKey, 32)
                    SecretKeySpec(transform.toByteArray(), AES.ALGORITHM)
                } else {
                    AES.EMPTY_SECRET_KEY
                }
            }
            else -> AES.EMPTY_SECRET_KEY
        }
    }

    fun genIvKey(ivKey: String?): IvParameterSpec {
        return if (ivKey != null) {
            val transform = genExactlyCharacterLengthKey(ivKey, 16)
            IvParameterSpec(transform.toByteArray())
        } else {
            AES.EMPTY_IV_SPEC
        }
    }
}

File: ios/Classes/CryptoHelper.swift
import Foundation
import CryptoSwift

class CryptoHelper {
    enum CryptoType {
        case encrypt
        case decrypt
    }
    struct Aes {
        static let algorithm = "AES"
        static let emptyIvSpec = Array<UInt8>(repeating: 0x00, count: 16)
        static let emptySecretKey = Array<UInt8>(repeating: 0x00, count: 32)
    }
    internal static let errValuePrivateKeyNull = "Value and privateKey must not be null or empty"
    internal static let errValuePrivateRequired = "Value and privateKey are required"
    internal static let errValidKey = "Invalid key or corrupted data"
    internal static let errIncorrectBlockSize = "Incorrect block size"
    internal static let errUnexpectedError = "Unexpected error"
    
    private func genExactlyCharacterLengthKey(key: String, length: Int) -> String {
        let bytes = Array(key.utf8)
        let hexString = bytes.map { String(format: "%02x", $0) }.joined()
        let hexLength = hexString.prefix(length).padding(toLength: length, withPad: "0", startingAt: 0)
        return String(hexLength)
    }
    
    func genSecretKey(algorithmType: String, secretKey: String?) -> Array<UInt8> {
        switch algorithmType {
        case CryptoHelper.Aes.algorithm:
            if let secretKey = secretKey {
                let transform = genExactlyCharacterLengthKey(key: secretKey, length: 32)
                return Array(transform.utf8)
            } else {
                return Aes.emptySecretKey
            }
        default:
            return Aes.emptySecretKey
        }
    }
    
    func genAESSecretKey(secretKey: String?) -> Array<UInt8> {
        if let secretKey = secretKey {
            let transform = genExactlyCharacterLengthKey(key: secretKey, length: 32)
            return Array(transform.utf8)
        } else {
            return Aes.emptySecretKey
        }
    }
    
    func genAESIvKey(ivKey: String?) -> Array<UInt8> {
        if let ivKey = ivKey {
            let transform = genExactlyCharacterLengthKey(key: ivKey, length: 16)
            return Array(transform.utf8)
        } else {
            return Aes.emptyIvSpec
        }
    }
}
